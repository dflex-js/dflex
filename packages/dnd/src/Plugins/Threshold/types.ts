/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { Rect } from "packages/core-instance/src/types";

export interface ThresholdPercentages {
  vertical: number;
  horizontal: number;
}

/**
 * Threshold in pixel relevant to input percentages.
 */
export interface ThresholdInPixels {
  x: number;
  y: number;
}

/**
 * Four directions threshold.
 */
export interface ThresholdMatrix {
  maxBottom: number;
  maxTop: number;
  maxLeft: number;
  maxRight: number;
}

export interface ThresholdInterface {
  thresholdPercentages: ThresholdPercentages;
  thresholdPixels: ThresholdInPixels;
  thresholdMatrix: ThresholdMatrix;
  getThresholdMatrix(
    top: number,
    left: number,
    height?: number,
    relativeToViewport?: boolean
  ): ThresholdMatrix;
  updateElementThresholdMatrix(
    elementRect: Rect,
    relativeToContainer: boolean,
    relativeToViewport?: boolean
  ): void;
}
