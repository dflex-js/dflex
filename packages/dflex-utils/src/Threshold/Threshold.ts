/* eslint-disable max-classes-per-file */
import { PointNum } from "../Point";

import type { Dimensions, Axis, Direction } from "../types";

import { AbstractBox, BoxBool, BoxNum, AbstractBoxRect } from "../Box";

export interface ThresholdPercentages {
  /** vertical threshold in percentage from 0-100 */
  vertical: number;

  /** horizontal threshold in percentage from 0-100 */
  horizontal: number;
}

class DFlexThreshold {
  readonly thresholds: Record<string, BoxNum>;

  private _pixels!: PointNum;

  private _percentages: ThresholdPercentages;

  isOut: Record<string, BoxBool>;

  constructor(percentages: ThresholdPercentages) {
    this._percentages = percentages;
    this.thresholds = {};
    this.isOut = {};
  }

  private _createPixels({ width, height }: Dimensions): void {
    const x = Math.round((this._percentages.horizontal * width) / 100);
    const y = Math.round((this._percentages.vertical * height) / 100);

    this._pixels = new PointNum(x, y);
  }

  /** Assign threshold property and create new instance for is out indicators */
  private _createThreshold(
    key: string,
    box: AbstractBox,
    isInner: boolean
  ): void {
    if (__DEV__) {
      if (this.thresholds[key] || this.isOut[key]) {
        throw new Error(`Threshold with key: ${key} already exists`);
      }
    }

    this.thresholds[key] = this._pixels.composeBox(box, isInner);
    console.log("ðŸš€ ~ file: Threshold.ts:51 ~ DFlexThreshold ~ box:", box);

    this.isOut[key] = new BoxBool(false, false, false, false);
  }

  private _setDepthThreshold(key: string, depth: number): void {
    const dp = `${depth}`;

    if (!this.thresholds[dp]) {
      this._createThreshold(dp, this.thresholds[key], false);

      return;
    }

    this.thresholds[depth].assignBiggestBox(this.thresholds[key]);
  }

  /**
   * Set the main threshold for the element based on the element's dimensions
   * and threshold types. For dragged and containers threshold type is outer
   * `isInner=false` and for the rest of the elements `isInner=true.`
   * Note: Duplicate threshold keys will throw an error.
   *
   * @param key
   * @param box
   * @param isInner
   */
  setMainThreshold(key: string, box: AbstractBoxRect, isInner: boolean): void {
    this._createPixels(box);

    this._createThreshold(key, box, isInner);
  }

  /**
   * Update existing threshold with new dimensions.
   *
   * @param key
   * @param rect
   * @param isInner
   */
  updateMainThreshold(key: string, rect: AbstractBox, isInner: boolean): void {
    if (__DEV__) {
      if (!this.thresholds[key]) {
        throw new Error(`Threshold ${key} does not exist.`);
      }
    }

    this.thresholds[key] = this._pixels.composeBox(rect, isInner);

    this.isOut[key].setFalsy();
  }

  getElmMainThreshold(rect: AbstractBox): BoxNum {
    return this._pixels.composeBox(rect, false);
  }

  /**
   * Assign outer threshold for the container. Along with another threshold
   * called insertion threshold which defines the area where the element can be
   * inserted during the migration taking into consideration the biggest hight
   * and width for the depth by using `unifiedContainerDimensions`. And create
   * accumulated depth threshold.
   *
   * @param SK
   * @param childDepth
   * @param containerRect
   * @param unifiedContainerDimensions
   */
  setContainerThreshold(
    SK: string,
    insertionLayerKey: string,
    childDepth: number,
    containerRect: AbstractBox,
    unifiedContainerDimensions: Dimensions
  ): void {
    // Regular threshold.
    this._createThreshold(SK, containerRect, false);

    const { top, left } = containerRect;
    const { height, width } = unifiedContainerDimensions;

    // Insertion threshold.
    this._createThreshold(
      insertionLayerKey,
      {
        left,
        top,
        right: left + width,
        bottom: top + height,
      },
      false
    );

    // Accumulated depth threshold. Accumulation based on insertion layer.
    this._setDepthThreshold(insertionLayerKey, childDepth);
  }

  /**
   * Check if the element is out of the threshold from given axis.
   *
   * Note: This method checks and sets so the result is always preserved.
   *
   *
   * @param axis
   * @param key
   * @param startingPos
   * @param endingPos
   * @returns
   */
  isOutThresholdByAxis(
    axis: Axis,
    key: string,
    startingPos: number,
    endingPos: number
  ): boolean {
    const { left, right, top, bottom } = this.thresholds[key];

    if (axis === "x") {
      this.isOut[key].setByAxis(axis, startingPos < left, endingPos > right);
    }

    if (axis === "y") {
      this.isOut[key].setByAxis(axis, startingPos < top, endingPos > bottom);
    }

    return this.isOut[key].isOneTruthyByAxis(axis);
  }

  /**
   * Check if the element is out of the threshold from one direction based only
   * on axis and element direction.
   *
   * Note: This method checks and sets so the result is always preserved
   *
   * @param axis
   * @param direction
   * @param key
   * @param startingPos
   * @param endingPos
   * @returns
   */
  isOutThresholdByDirection(
    axis: Axis,
    direction: Direction,
    key: string,
    startingPos: number,
    endingPos: number
  ): boolean {
    const { left, right, top, bottom } = this.thresholds[key];

    const is =
      axis === "x"
        ? direction === -1
          ? startingPos < left
          : endingPos > right
        : direction === -1
        ? startingPos < top
        : endingPos > bottom;

    this.isOut[key].setOne(axis, direction, is);

    return is;
  }

  isOutThreshold(key: string, box: BoxNum, isPreserveResult: boolean): boolean {
    const thresholdBox = this.thresholds[key];

    return box.isOutThreshold(
      thresholdBox,
      isPreserveResult ? this.isOut[key] : undefined
    );
  }

  destroy(): void {
    Object.keys(this.thresholds).forEach((key) => {
      delete this.thresholds[key];
    });
    Object.keys(this.isOut).forEach((key) => {
      delete this.isOut[key];
    });
  }
}

export default DFlexThreshold;
