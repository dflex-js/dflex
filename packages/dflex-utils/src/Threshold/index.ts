export { default as DFlexThreshold } from "./Threshold";

export type { ThresholdPercentages } from "./Threshold";
