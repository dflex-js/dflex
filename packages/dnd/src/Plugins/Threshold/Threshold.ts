/* eslint-disable max-classes-per-file */
import { AxesCoordinates, AxesCoordinatesInterface, Rect } from "@dflex/utils";
import type {
  ThresholdInterface,
  ThresholdPointInterface,
  ThresholdCoordinate,
  ThresholdPercentages,
} from "./types";

class ThresholdPoint implements ThresholdPointInterface {
  max: number;

  min: number;

  constructor(max: number, min: number) {
    this.max = max;
    this.min = min;
  }
}

class Threshold implements ThresholdInterface {
  percentages: ThresholdInterface["percentages"];

  private pixels!: AxesCoordinatesInterface;

  main!: ThresholdCoordinate;

  constructor(
    percentages: ThresholdPercentages,
    rect: Rect,
    { isContainer }: { isContainer: boolean }
  ) {
    this.percentages = percentages;

    this.setMainThreshold(rect, isContainer);
  }

  private setPixels({ width, height }: Rect) {
    const x = Math.round((this.percentages.horizontal * width) / 100);

    const y = Math.round((this.percentages.vertical * height) / 100);

    this.pixels = new AxesCoordinates(x, y);
  }

  getThreshold(rect: Rect, isContainer: boolean) {
    const { top, left, height } = rect;

    const { x, y } = this.pixels;

    const leftThresholdPoint = new ThresholdPoint(left - x, left + x);

    const topThresholdPoint = new ThresholdPoint(
      top - y,
      isContainer ? Math.abs(height - y) : top + y
    );

    return {
      left: leftThresholdPoint,
      top: topThresholdPoint,
    };
  }

  setMainThreshold(rect: Rect, isContainer: boolean) {
    this.setPixels(rect);

    this.main = this.getThreshold(rect, isContainer);
  }
}

export default Threshold;
