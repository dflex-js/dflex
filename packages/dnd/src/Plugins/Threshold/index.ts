export type { ThresholdInterface, ThresholdMatrix } from "./types";

export { default } from "./Threshold";
