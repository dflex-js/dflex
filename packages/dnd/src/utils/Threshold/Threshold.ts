/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type { ThresholdInterface, ThresholdMatrix } from "./types";

class Threshold implements ThresholdInterface {
  thresholdPercentages: ThresholdInterface["thresholdPercentages"];

  thresholdPixels!: ThresholdInterface["thresholdPixels"];

  thresholdMatrix!: ThresholdInterface["thresholdMatrix"];

  constructor(thresholdInput: ThresholdInterface["thresholdPercentages"]) {
    this.thresholdPercentages = thresholdInput;
  }

  getThresholdMatrix(
    top: number,
    left: number,
    height?: number
  ): ThresholdMatrix {
    const { x, y } = this.thresholdPixels;

    /**
     * When going up, currentTop decreases (-vertical).
     */
    const maxTop = top - y;

    /**
     * When going left, currentLeft decreases (-horizontal).
     */
    const maxLeft = left - x;

    /**
     * When going right, currentLeft increases (+horizontal) with droppable
     * taking into considerations (+ horizontal).
     */
    const maxRight = left + x;

    /**
     * If height, the threshold is relative to the container. Otherwise, it's
     * relative to the element.
     */
    const maxBottom = height ? height - y : top + y;

    return {
      maxBottom,
      maxTop,
      maxLeft,
      maxRight,
    };
  }

  updateElementThresholdMatrix(
    width: number,
    height: number,
    left: number,
    top: number
  ) {
    const x = Math.round((this.thresholdPercentages.horizontal * width) / 100);

    const y = Math.round((this.thresholdPercentages.vertical * height) / 100);

    this.thresholdPixels = { x, y };

    this.thresholdMatrix = this.getThresholdMatrix(top, left);
  }
}

export default Threshold;
