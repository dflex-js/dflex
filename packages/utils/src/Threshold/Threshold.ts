import {
  AxesCoordinates,
  AxesCoordinatesInterface,
  AxesFourCoordinatesBool,
} from "../AxesCoordinates";

import { RectDimensions, RectBoundaries } from "../types";
import type {
  ThresholdInterface,
  ThresholdPercentages,
  ThresholdsStore,
  LayoutPositionStatus,
} from "./types";

class Threshold implements ThresholdInterface {
  #pixels!: AxesCoordinatesInterface;

  #percentages: ThresholdPercentages;

  thresholds: ThresholdsStore;

  isOut: LayoutPositionStatus;

  constructor(percentages: ThresholdPercentages) {
    this.#percentages = percentages;

    this.thresholds = {};
    this.isOut = {};
  }

  #setPixels({ width, height }: RectDimensions) {
    const x = Math.round((this.#percentages.horizontal * width) / 100);
    const y = Math.round((this.#percentages.vertical * height) / 100);

    this.#pixels = new AxesCoordinates(x, y);
  }

  #initIndicators(key: string) {
    if (!this.isOut[key]) {
      this.isOut[key] = new AxesFourCoordinatesBool();
    } else {
      this.isOut[key].reset();
    }
  }

  private getScrollThreshold(rect: RectDimensions) {
    const { top, left, height, width } = rect;

    const { x, y } = this.#pixels;

    return {
      left: Math.abs(left - x),
      right: left - x + width,
      top: Math.abs(top - y),
      bottom: height - y,
    };
  }

  setScrollThreshold(key: string, rect: RectDimensions) {
    this.#setPixels(rect);

    this.thresholds[key] = this.getScrollThreshold(rect);

    if (!this.isOut[key]) {
      this.isOut[key] = new AxesFourCoordinatesBool();
    } else {
      this.isOut[key].reset();
    }
  }

  private getThreshold(rect: RectBoundaries) {
    const { top, left, bottom, right } = rect;

    const { x, y } = this.#pixels;

    return {
      left: left - x,
      right: right + x,
      top: top - y,
      bottom: bottom + y,
    };
  }

  setMainThreshold(key: string, rect: RectDimensions) {
    this.#setPixels(rect);

    const { top, left, height, width } = rect;

    this.thresholds[key] = this.getThreshold({
      top,
      left,
      bottom: top + height,
      right: left + width,
    });

    this.#initIndicators(key);
  }

  setContainerThreshold(key: string, rect: RectBoundaries) {
    this.thresholds[key] = this.getThreshold(rect);
    this.#initIndicators(key);
  }

  isOutThresholdH(key: string, XLeft: number, XRight: number) {
    const { left, right } = this.thresholds[key];

    this.isOut[key].setOutX(XLeft < left, XRight > right);

    return this.isOut[key].isOutX();
  }

  isOutThresholdV(key: string, YTop: number, YBottom: number) {
    const { top, bottom } = this.thresholds[key];

    this.isOut[key].setOutY(YTop < top, YBottom > bottom);

    return this.isOut[key].isOutY();
  }
}

export default Threshold;
