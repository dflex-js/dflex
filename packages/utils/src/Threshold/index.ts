export { default as Threshold } from "./Threshold";

export type {
  ThresholdInterface,
  ThresholdCoordinate,
  ThresholdPercentages,
} from "./types";
