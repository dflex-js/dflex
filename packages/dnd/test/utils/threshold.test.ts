/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Threshold from "../../src/utils/Threshold";

describe("Threshold", () => {
  it("Has the right instance", () => {
    const thresholdPercentages = { vertical: 60, horizontal: 60 };

    const threshold = new Threshold(thresholdPercentages, 200, 100);

    expect(threshold).toMatchSnapshot();
  });
});
