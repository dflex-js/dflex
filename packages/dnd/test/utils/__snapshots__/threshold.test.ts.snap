// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Threshold Has the right instance 1`] = `
Threshold {
  "thresholdPercentages": Object {
    "horizontal": 60,
    "vertical": 60,
  },
}
`;
